internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
+				// original context so operations aren't immediately cancelled.
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
+
+				return
+			case <-c.debounceChan.C():
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
 			}
-		}()
+		}
+	}()
+
+	return nil
+}
 
+func (c *BatchCompleter) handleBatch(ctx context.Context) error {
+	var setStateBatch map[int64]*batchCompleterSetState
+	func() {
+		c.setStateParamsMu.Lock()
+		defer c.setStateParamsMu.Unlock()
+
+		setStateBatch = c.setStateParams
+
+		// Don't bother resetting the map if there's nothing to process,
+		// allowing the completer to idle efficiently.
+		if len(setStateBatch) > 0 {
+			c.setStateParams = make(map[int64]*batchCompleterSetState)
+		} else {
+			// Set nil to avoid a data race below in case the map is set as a
+			// new job comes in.
+			setStateBatch = nil
+		}
+	}()
+
+	if len(setStateBatch) < 1 {
 		return nil
-	})
+	}
+
+	// Insert a sub-batch with retries. Also helps reduce visual noise and
+	// increase readability of loop below.
+	insertSubBatch := func(batchID []int64, batchFinalizedAt []time.Time) ([]*rivertype.JobRow, error) {
+		return withRetries(ctx, &c.BaseService, func(ctx context.Context) ([]*rivertype.JobRow, error) {
+			return c.exec.JobSetCompleteIfRunningMany(ctx, &riverdriver.JobSetCompleteIfRunningManyParams{
+				ID:          batchID,
+				FinalizedAt: batchFinalizedAt,
+			})
+		})
+	}
+
+	// This could be written more simply using multiple `sliceutil.Map`s, but
+	// it's done this way to allocate as few new slices as necessary.
+	mapIDsAndFinalizedAt := func(setStateBatch map[int64]*batchCompleterSetState) ([]int64, []time.Time) {
+		var (
+			batchIDs         = make([]int64, len(setStateBatch))
+			batchFinalizedAt = make([]time.Time, len(setStateBatch))
+			i                int
+		)
+		for _, setState := range setStateBatch {
+			batchIDs[i] = setState.Params.ID
+			batchFinalizedAt[i] = *setState.Params.FinalizedAt
+			i++
+		}
+		return batchIDs, batchFinalizedAt
+	}
+
+	// Tease apart enormous batches into sub-batches.
+	//
+	// All the code below is concerned with doing that, with a fast loop that
+	// doesn't allocate any additional memory in case the entire batch is
+	// smaller than the sub-batch maximum size (which will be the common case).
+	const oneOperationMax = 2_000
+
+	var (
+		batchID, batchFinalizedAt = mapIDsAndFinalizedAt(setStateBatch)
+		jobRows                   []*rivertype.JobRow
+	)
+	if len(setStateBatch) > oneOperationMax {
+		jobRows = make([]*rivertype.JobRow, 0, len(setStateBatch))
+		for i := 0; i < len(setStateBatch); i += oneOperationMax {
+			endIndex := min(i+oneOperationMax, len(batchID)) // beginning of next sub-batch or end of slice
+			jobRowsSubBatch, err := insertSubBatch(batchID[i:endIndex], batchFinalizedAt[i:endIndex])
+			if err != nil {
+				return err
+			}
```

Comment:
so if a single sub batch fails after 4 attempts, we immediately return an error and give up on all remaining jobs in the batch? That feels a bit risky 🤔 

Also is there any kind of backpressure to stop the client from fetching more jobs if the completer is filling up and can't complete all the jobs it already has lined up?

Reply:
So I ended up putting in a backpressure system, wherein the batch completer blocks are receiving more jobs if the backlog is too large, and added a batch completer test to try and exercise it specifically.

Hitting maximum backlog produces a warning log line, and I've found this to be very helpful for benchmarking purposes. Previously a backlogged completer was swept under the rug because infinite backlog was allowed, so a maximum backlog combined with backpressure forces the benchmark to pause until jobs can clear.

In terms of giving up after four attempts: I think unfortunately there's only so much we can do there. If the completer's failed four times, then likely the database is in a totally degenerate state or the client's lost its database connection, either of which are going to bubble up and create additional carnage in places other than completer. This is better now with the backpressure system though -- if the completer's having trouble getting a completion through, jobs accumulate until the backlog gets hit, and the whole client gets effectively paused until jobs can start completing again. I've observed this in my benchmarks from time to time as my Postgres degrades in odd ways.